keps/sig-api-machinery/3945-crd-conversion-with-cel/README.md

```diff
@@ -0,0 +1,883 @@
+
+Current the conversion strategy in CRD only supports 'none' and 'webhook'. This KEP proposes to add a new strategy type
+'ConversionRules' which uses CEL for CRD version conversion.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  # name must match the spec fields below, and be in the form: <plural>.<group>
+  name: crontabs.example.com
+spec:
+  ...
+  conversion:
+    # a ConversionRules strategy instruct API server to use CEL for any conversion between custom resources.
+    strategy: ConversionRules
+    conversionRules:
+      - from: v1 # from is the source version.
+        to: v2 # to is the target version.
+        rules: # rules contains field-specific conversion expressions.
+          # field is a JSONPath expression to the field in the originating version of the object, relative to
+          #  its root
+          - field: .spec.firstName 
+            # destination is a JSONPath expression to the
+            # field in the target version of the object, relative
+            # to its root
+            destination: .spec.name.first
+            # transformation is an optional CEL expression used to execute user-specified rules to transform the
+            # originating field -- identified by 'self' -- to the destination field.
+            transformation: self
+...
+```
+
+A new conversion strategy `ConversionRules` is added to instruct API server to use CEL based on following conversionRules
+for any conversion between custom resources.
+
+conversionRules is to specify rules to convert between different API versions:
+- `from` is the source version. 
+- `to` is the target version.
+- `rules` contains field-specific conversion expressions.
+  - `field` is a JSONPath expression to the field in the originating version of the object, relative to its root.
+  - `destination` is a JSONPath expression to the field in the target version of the object, relative to its root.
+  - `transformation` is an optional CEL expression used to execute user-specified rules to transform the originating field.
+    if absent, behaves as-if the rule is set to "self" (CEL expression that assigns the value of the source to the target).
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1 - rename a field
+
+#### Story 2 - relocate a field
+
+#### Story 3 - convert a annotation or label to a field
+
+#### Story 4 - convert a singular field to plural
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
Quick reminder that conversion is performed on the responses to LIST requests, so if you've got a large list the amount of conversions can also be large. Potential risks might be hidden here?